internal/retry/backoff.go

```diff
@@ -219,32 +223,51 @@ func (b *Backoffer) String() string {
 	return fmt.Sprintf(" backoff(%dms %v)", b.totalSleep, b.configs)
 }
 
+// copyMapWithoutRecursive is only used to deep copy map fields in the Backoffer type.
+func copyMapWithoutRecursive(srcMap map[string]int) map[string]int {
+	result := map[string]int{}
+	for k, v := range srcMap {
+		result[k] = v
+	}
+	return result
+}
+
 // Clone creates a new Backoffer which keeps current Backoffer's sleep time and errors, and shares
 // current Backoffer's context.
+// Some fields like `configs` and `vars` are concurrently used by all the backoffers in different threads,
+// try not to modify the referenced content directly.
 func (b *Backoffer) Clone() *Backoffer {
 	return &Backoffer{
-		ctx:           b.ctx,
-		maxSleep:      b.maxSleep,
-		totalSleep:    b.totalSleep,
-		excludedSleep: b.excludedSleep,
-		errors:        b.errors,
-		vars:          b.vars,
-		parent:        b.parent,
+		ctx:            b.ctx,
+		maxSleep:       b.maxSleep,
+		totalSleep:     b.totalSleep,
+		excludedSleep:  b.excludedSleep,
+		vars:           b.vars,
+		errors:         append([]error{}, b.errors...),
+		configs:        append([]*Config{}, b.configs...),
```

Comment:
```suggestion
		errors:         append(make([]error, len(b.errors)), b.errors...),
		configs:        append(make([]*Config, len(b.errors)), b.configs...),
```

Reply:
Could `append(make([]*Config, 0, len(b.errors)), b.configs...)` be a faster way? I'm not quite sure about it.

Reply:
Oh, yes. Should be `make([]*Config, 0, len(b.errors))`. It should be faster, I think.

Reply:
I doubt it too. I try these two, and they perform the same.

```go
func BenchmarkAppend1(b *testing.B) {
	source := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
	for i := 0; i < b.N; i++ {
		_ = append([]int{}, source...)
	}
}

func BenchmarkAppend2(b *testing.B) {
	source := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
	for i := 0; i < b.N; i++ {
		_ = append(make([]int, 0, len(source)), source...)
	}
}
```

```
BenchmarkAppend1-24     49070558                40.73 ns/op           80 B/op          1 allocs/op
BenchmarkAppend2-24     38036925                39.76 ns/op           80 B/op          1 allocs/op
```

Reply:
>  It should be faster, I think.

To be more accurate, it can't be slower.

Reply:
I tried some benchmarks and found it's not faster because it still use `append()` to copy data. If we use `copy()` here, I think it should be faster.